src/TAMSEquationSystem.C

```diff
@@ -0,0 +1,615 @@
+//-------- register_interior_algorithm -------------------------------------
+//--------------------------------------------------------------------------
+void
+TAMSEquationSystem::register_interior_algorithm(stk::mesh::Part* part)
+{
+
+  // types of algorithms
+  const AlgorithmType algType = INTERIOR;
+
+  // metric tensor algorithm
+  if (NULL == metricTensorAlgDriver_)
+    metricTensorAlgDriver_ = new AlgorithmDriver(realm_);
+
+  std::map<AlgorithmType, Algorithm*>::iterator itmt =
+    metricTensorAlgDriver_->algMap_.find(algType);
+
+  if (itmt == metricTensorAlgDriver_->algMap_.end()) {
+    ComputeMetricTensorNodeAlgorithm* metricTensorAlg =
+      new ComputeMetricTensorNodeAlgorithm(realm_, part);
+    metricTensorAlgDriver_->algMap_[algType] = metricTensorAlg;
+  } else {
+    itmt->second->partVec_.push_back(part);
+  }
+
+  // averaging algorithm
+  if (NULL == averagingAlgDriver_)
+    averagingAlgDriver_ = new AlgorithmDriver(realm_);
+
+  std::map<AlgorithmType, Algorithm*>::iterator itav =
+    averagingAlgDriver_->algMap_.find(algType);
+
+  if (itav == averagingAlgDriver_->algMap_.end()) {
+    Algorithm* theAlg = NULL;
+    switch (turbulenceModel_) {
+    case SST_TAMS:
+      theAlg = new ComputeSSTTAMSAveragesNodeAlgorithm(realm_, part);
+      break;
+    default:
+      throw std::runtime_error("TAMSEquationSystem: non-supported turb model");
+    }
+    averagingAlgDriver_->algMap_[algType] = theAlg;
+  } else {
+    itav->second->partVec_.push_back(part);
+  }
+
+  // avgMdot algorithm
+  if (NULL == avgMdotAlgDriver_)
+    avgMdotAlgDriver_ = new AlgorithmDriver(realm_);
+
+  if (realm_.realmUsesEdges_) {
+    std::map<AlgorithmType, Algorithm*>::iterator itmd =
+      avgMdotAlgDriver_->algMap_.find(algType);
+
+    if (itmd == avgMdotAlgDriver_->algMap_.end()) {
+      ComputeTAMSAvgMdotEdgeAlgorithm* avgMdotEdgeAlg =
+        new ComputeTAMSAvgMdotEdgeAlgorithm(realm_, part);
+      avgMdotAlgDriver_->algMap_[algType] = avgMdotEdgeAlg;
+    } else {
+      itmd->second->partVec_.push_back(part);
+    }
+  } else {
+    std::map<AlgorithmType, Algorithm*>::iterator itmd =
+      avgMdotAlgDriver_->algMap_.find(algType);
+
+    if (itmd == avgMdotAlgDriver_->algMap_.end()) {
+      ComputeTAMSAvgMdotElemAlgorithm* avgMdotAlg =
+        new ComputeTAMSAvgMdotElemAlgorithm(realm_, part);
+      avgMdotAlgDriver_->algMap_[algType] = avgMdotAlg;
+    } else {
+      itmd->second->partVec_.push_back(part);
+    }
+  }
+
+  // FIXME: tvisc needed for TAMS update, but is updated in LowMach...
+  //        Perhaps there is a way to call tvisc from LowMach here?
+  std::map<AlgorithmType, Algorithm*>::iterator it_tv =
+    tviscAlgDriver_->algMap_.find(algType);
+  if (it_tv == tviscAlgDriver_->algMap_.end()) {
+    Algorithm* theAlg = NULL;
+    switch (realm_.solutionOptions_->turbulenceModel_) {
+      case SST_TAMS:
+        theAlg = new TurbViscSSTTAMSAlgorithm(realm_, part);
```

Comment:
There are two versions of `tvisc` floating around? If LowMach updates this, why is this needed again here?

Reply:
The issue I was trying to resolve was that the solver does LowMach / K, Omega / TAMS and in TAMS we use tvisc during part of the average update, but since tvisc was calculated before the K and Omega updates it is inconsistent with their values at that point.

If you can think of a better way to handle this, I would certainly be in board with that.